File: packages/kerf-api/src/kerf_api/tools/project_layers.py
"""
Project-level layer + display-mode LLM tools.

All tools operate on the project's `.canvas.json` file (kind='canvas'),
creating it at the project root if it does not yet exist.
"""

import json
import re
import uuid
from tools.registry import ToolSpec, err_payload, ok_payload, register
from tools.context import ProjectCtx


# ---------------------------------------------------------------------------
# Default canvas (mirrors projectLayers.js defaultCanvas())
# ---------------------------------------------------------------------------

def _default_canvas() -> dict:
    return {
        "version": 1,
        "layers": [
            {
                "id": "L01",
                "name": "Geometry",
                "visible": True,
                "color": "#ffffff",
                "linetype": "continuous",
                "material_id": None,
                "locked": False,
            }
        ],
        "display_modes": [
            {"id": "shaded",    "name": "Shaded",    "wireframe": False, "edges": True,  "shadows": False, "transparency": 1.0,  "background_color": "#1a1a1a"},
            {"id": "wireframe", "name": "Wireframe", "wireframe": True,  "edges": True},
            {"id": "technical", "name": "Technical", "wireframe": False, "edges": True,  "silhouette": True, "shadows": False},
            {"id": "rendered",  "name": "Rendered",  "wireframe": False, "edges": False, "shadows": True,  "transparency": 0.95},
        ],
        "active_display_mode": "shaded",
        "active_layer": "L01",
    }

_HEX_RE = re.compile(r'^#[0-9a-fA-F]{3}([0-9a-fA-F]{3})?$')


def _next_layer_id(layers: list[dict]) -> str:
    existing = {l["id"] for l in layers}
    n = len(layers) + 1
    while True:
        cid = f"L{n:02d}"
        if cid not in existing:
            return cid
        n += 1


# ---------------------------------------------------------------------------
# DB helpers
# ---------------------------------------------------------------------------

async def _load_canvas(ctx: ProjectCtx) -> tuple[uuid.UUID | None, dict]:
    """Return (file_id | None, canvas_dict). file_id is None if not yet saved."""
    row = await ctx.pool.fetchrow(
        "SELECT id, content FROM files "
        "WHERE project_id = $1 AND name = '.canvas.json' AND deleted_at IS NULL "
        "LIMIT 1",
        ctx.project_id,
    )
    if not row:
        return None, _default_canvas()
    try:
        canvas = json.loads(row["content"] or "{}")
        if not isinstance(canvas, dict) or "layers" not in canvas:
            canvas = _default_canvas()
    except Exception:
        canvas = _default_canvas()
    return row["id"], canvas


async def _save_canvas(ctx: ProjectCtx, file_id: uuid.UUID | None, canvas: dict) -> uuid.UUID:
    """Upsert the .canvas.json file; return its file_id."""
    content = json.dumps(canvas, indent=2)
    if file_id is None:
        new_id = uuid.uuid4()
        await ctx.pool.execute(
            "INSERT INTO files(id, project_id, parent_id, name, kind, content) "
            "VALUES ($1, $2, NULL, '.canvas.json', 'canvas', $3)",
            new_id, ctx.project_id, content,
        )
        return new_id
    else:
        await ctx.pool.execute(
            "UPDATE files SET content = $1, updated_at = NOW() WHERE id = $2",
            content, file_id,
        )
        return file_id


# ---------------------------------------------------------------------------
# Tool: create_layer
# ---------------------------------------------------------------------------

create_layer_spec = ToolSpec(
    name="create_layer",
    description=(
        "Create a new project layer. "
        "color must be a hex string like '#ff0000'. "
        "If a layer with the same name already exists the call is a no-op."
    ),
    input_schema={
        "type": "object",
        "properties": {
            "project_id": {"type": "string"},
            "name":       {"type": "string"},
            "color":      {"type": "string", "default": "#aaaaaa"},
            "linetype":   {"type": "string", "default": "continuous"},
            "locked":     {"type": "boolean", "default": False},
        },
        "required": ["project_id", "name"],
    },
)


@register(create_layer_spec, write=True)
async def run_create_layer(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    name = a.get("name", "").strip()
    if not name:
        return err_payload("name is required", "BAD_ARGS")

    color    = a.get("color", "#aaaaaa")
    linetype = a.get("linetype", "continuous")
    locked   = bool(a.get("locked", False))

    if not _HEX_RE.match(color):
        return err_payload(f"invalid color: {color}", "BAD_ARGS")

    file_id, canvas = await _load_canvas(ctx)
    if any(l["name"] == name for l in canvas["layers"]):
        return ok_payload({"message": f"layer '{name}' already exists — no-op"})

    new_id = _next_layer_id(canvas["layers"])
    canvas["layers"].append({
        "id": new_id, "name": name, "visible": True,
        "color": color, "linetype": linetype,
        "material_id": None, "locked": locked,
    })
    await _save_canvas(ctx, file_id, canvas)
    return ok_payload({"layer_id": new_id, "message": f"created layer '{name}' as {new_id}"})


# ---------------------------------------------------------------------------
# Tool: delete_layer
# ---------------------------------------------------------------------------

delete_layer_spec = ToolSpec(
    name="delete_layer",
    description="Delete a project layer by id. Refuses if it is the last layer.",
    input_schema={
        "type": "object",
        "properties": {
            "project_id": {"type": "string"},
            "layer_id":   {"type": "string"},
        },
        "required": ["project_id", "layer_id"],
    },
)


@register(delete_layer_spec, write=True)
async def run_delete_layer(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    layer_id = a.get("layer_id", "").strip()
    if not layer_id:
        return err_payload("layer_id is required", "BAD_ARGS")

    file_id, canvas = await _load_canvas(ctx)
    filtered = [l for l in canvas["layers"] if l["id"] != layer_id]
    if len(filtered) == len(canvas["layers"]):
        return ok_payload({"message": f"layer '{layer_id}' not found — no-op"})
    if len(filtered) == 0:
        return err_payload("cannot remove the last layer", "PROTECTED")

    canvas["layers"] = filtered
    if canvas.get("active_layer") == layer_id:
        canvas["active_layer"] = filtered[0]["id"]

    await _save_canvas(ctx, file_id, canvas)
    return ok_payload({"message": f"deleted layer '{layer_id}'"})


# ---------------------------------------------------------------------------
# Tool: set_layer_visibility
# ---------------------------------------------------------------------------

set_layer_visibility_spec = ToolSpec(
    name="set_project_layer_visibility",
    description="Show or hide a project layer.",
    input_schema={
        "type": "object",
        "properties": {
            "project_id": {"type": "string"},
            "layer_id":   {"type": "string"},
            "visible":    {"type": "boolean"},
        },
        "required": ["project_id", "layer_id", "visible"],
    },
)


@register(set_layer_visibility_spec, write=True)
async def run_set_layer_visibility(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    layer_id = a.get("layer_id", "").strip()
    visible  = bool(a.get("visible", True))

    file_id, canvas = await _load_canvas(ctx)
    found = False
    for l in canvas["layers"]:
        if l["id"] == layer_id:
            l["visible"] = visible
            found = True
            break
    if not found:
        return err_payload(f"layer '{layer_id}' not found", "NOT_FOUND")

    await _save_canvas(ctx, file_id, canvas)
    return ok_payload({"message": f"set '{layer_id}' visible={visible}"})


# ---------------------------------------------------------------------------
# Tool: set_layer_color
# ---------------------------------------------------------------------------

set_layer_color_spec = ToolSpec(
    name="set_project_layer_color",
    description="Set the hex color of a project layer.",
    input_schema={
        "type": "object",
        "properties": {
            "project_id": {"type": "string"},
            "layer_id":   {"type": "string"},
            "color":      {"type": "string", "description": "Hex color e.g. '#ff0000'"},
        },
        "required": ["project_id", "layer_id", "color"],
    },
)


@register(set_layer_color_spec, write=True)
async def run_set_layer_color(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    layer_id = a.get("layer_id", "").strip()
    color    = a.get("color", "").strip()

    if not _HEX_RE.match(color):
        return err_payload(f"invalid color: {color}", "BAD_ARGS")

    file_id, canvas = await _load_canvas(ctx)
    found = False
    for l in canvas["layers"]:
        if l["id"] == layer_id:
            l["color"] = color
            found = True
            break
    if not found:
        return err_payload(f"layer '{layer_id}' not found", "NOT_FOUND")

    await _save_canvas(ctx, file_id, canvas)
    return ok_payload({"message": f"set '{layer_id}' color={color}"})


# ---------------------------------------------------------------------------
# Tool: assign_to_layer
# ---------------------------------------------------------------------------

assign_to_layer_spec = ToolSpec(
    name="assign_file_to_layer",
    description=(
        "Record which layer a project file belongs to. "
        "Stores the layer_id in the file's metadata JSON under key 'layer_id'. "
        "The file must belong to the same project."
    ),
    input_schema={
        "type": "object",
        "properties": {
            "project_id": {"type": "string"},
            "file_id":    {"type": "string", "description": "UUID of the file to assign"},
            "layer_id":   {"type": "string"},
        },
        "required": ["project_id", "file_id", "layer_id"],
    },
)


@register(assign_to_layer_spec, write=True)
async def run_assign_to_layer(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    raw_file_id = a.get("file_id", "").strip()
    layer_id    = a.get("layer_id", "").strip()

    if not raw_file_id or not layer_id:
        return err_payload("file_id and layer_id are required", "BAD_ARGS")

    try:
        target_file_id = uuid.UUID(raw_file_id)
    except ValueError:
        return err_payload(f"invalid file_id UUID: {raw_file_id}", "BAD_ARGS")

    # Verify the layer exists in canvas.
    _, canvas = await _load_canvas(ctx)
    if not any(l["id"] == layer_id for l in canvas["layers"]):
        return err_payload(f"layer '{layer_id}' not found in canvas", "NOT_FOUND")

    # Verify file belongs to project.
    row = await ctx.pool.fetchrow(
        "SELECT id, metadata FROM files WHERE id = $1 AND project_id = $2 AND deleted_at IS NULL",
        target_file_id, ctx.project_id,
    )
    if not row:
        return err_payload(f"file '{raw_file_id}' not found in project", "NOT_FOUND")

    existing_meta = {}
    if row["metadata"]:
        try:
            existing_meta = json.loads(row["metadata"])
        except Exception:
            pass
    existing_meta["layer_id"] = layer_id

    await ctx.pool.execute(
        "UPDATE files SET metadata = $1, updated_at = NOW() WHERE id = $2",
        json.dumps(existing_meta), target_file_id,
    )
    return ok_payload({"message": f"assigned file '{raw_file_id}' to layer '{layer_id}'"})


# ---------------------------------------------------------------------------
# Tool: switch_display_mode
# ---------------------------------------------------------------------------

switch_display_mode_spec = ToolSpec(
    name="switch_display_mode",
    description=(
        "Set the active display mode for the project's 3D viewport. "
        "Built-in modes: shaded | wireframe | technical | rendered."
    ),
    input_schema={
        "type": "object",
        "properties": {
            "project_id": {"type": "string"},
            "mode_id":    {"type": "string", "description": "Display mode id, e.g. 'wireframe'"},
        },
        "required": ["project_id", "mode_id"],
    },
)


@register(switch_display_mode_spec, write=True)
async def run_switch_display_mode(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    mode_id = a.get("mode_id", "").strip()
    if not mode_id:
        return err_payload("mode_id is required", "BAD_ARGS")

    file_id, canvas = await _load_canvas(ctx)
    if not any(m["id"] == mode_id for m in canvas["display_modes"]):
        valid = [m["id"] for m in canvas["display_modes"]]
        return err_payload(f"mode '{mode_id}' not found; valid: {valid}", "NOT_FOUND")

    canvas["active_display_mode"] = mode_id
    await _save_canvas(ctx, file_id, canvas)
    return ok_payload({"message": f"active display mode set to '{mode_id}'"})

File: web/src/lib/projectLayers.ts
/**
 * Project-level layers + display-mode helpers.
 * Operates on a plain `.canvas.json` object — no side effects.
 */

export interface CanvasLayer {
  id: string
  name: string
  visible: boolean
  color: string
  linetype: string
  material_id: string | null
  locked: boolean
}

export interface CanvasDisplayMode {
  id: string
  name: string
  wireframe: boolean
  edges: boolean
  shadows?: boolean
  transparency?: number
  background_color?: string
  silhouette?: boolean
}

export interface Canvas {
  version: 1
  layers: CanvasLayer[]
  display_modes: CanvasDisplayMode[]
  active_display_mode: string
  active_layer: string
}

const HEX_RE = /^#[0-9a-fA-F]{3}([0-9a-fA-F]{3})?$/

function _nextId(canvas: Canvas): string {
  const existing = new Set(canvas.layers.map((l) => l.id))
  let n = canvas.layers.length + 1
  while (existing.has(`L${String(n).padStart(2, '0')}`)) n++
  return `L${String(n).padStart(2, '0')}`
}

/** Return a fresh canvas with sensible defaults. */
export function defaultCanvas(): Canvas {
  return {
    version: 1,
    layers: [
      {
        id: 'L01',
        name: 'Geometry',
        visible: true,
        color: '#ffffff',
        linetype: 'continuous',
        material_id: null,
        locked: false,
      },
    ],
    display_modes: [
      { id: 'shaded',    name: 'Shaded',    wireframe: false, edges: true,  shadows: false, transparency: 1.0,  background_color: '#1a1a1a' },
      { id: 'wireframe', name: 'Wireframe', wireframe: true,  edges: true },
      { id: 'technical', name: 'Technical', wireframe: false, edges: true,  silhouette: true, shadows: false },
      { id: 'rendered',  name: 'Rendered',  wireframe: false, edges: false, shadows: true,  transparency: 0.95 },
    ],
    active_display_mode: 'shaded',
    active_layer: 'L01',
  }
}

/**
 * Add a new layer. Auto-assigns `id`. Returns new canvas (immutable).
 */
export function addLayer(canvas: Canvas, opts: { name: string; color?: string; linetype?: string; material_id?: string | null; locked?: boolean }): Canvas {
  const { name, color = '#ffffff', linetype = 'continuous', material_id = null, locked = false } = opts
  if (!name || !name.trim()) throw new Error('layer name required')
  const id = _nextId(canvas)
  const layer: CanvasLayer = { id, name: name.trim(), visible: true, color, linetype, material_id, locked }
  return { ...canvas, layers: [...canvas.layers, layer] }
}

/**
 * Remove layer by id. Files that were on the removed layer are implicitly
 * reassigned to `active_layer` by the caller; this function just removes the
 * entry and updates active_layer if it pointed to the removed one.
 */
export function removeLayer(canvas: Canvas, layer_id: string): Canvas {
  const filtered = canvas.layers.filter((l) => l.id !== layer_id)
  if (filtered.length === canvas.layers.length) return canvas // not found — no-op
  if (filtered.length === 0) throw new Error('cannot remove the last layer')
  let active = canvas.active_layer
  if (active === layer_id) active = filtered[0].id
  return { ...canvas, layers: filtered, active_layer: active }
}

export function setLayerVisibility(canvas: Canvas, layer_id: string, visible: boolean): Canvas {
  return {
    ...canvas,
    layers: canvas.layers.map((l) => l.id === layer_id ? { ...l, visible: Boolean(visible) } : l),
  }
}

export function setLayerColor(canvas: Canvas, layer_id: string, color: string): Canvas {
  if (!HEX_RE.test(color)) throw new Error(`invalid hex color: ${color}`)
  return {
    ...canvas,
    layers: canvas.layers.map((l) => l.id === layer_id ? { ...l, color } : l),
  }
}

export function setActiveLayer(canvas: Canvas, layer_id: string): Canvas {
  if (!canvas.layers.find((l) => l.id === layer_id)) throw new Error(`layer ${layer_id} not found`)
  return { ...canvas, active_layer: layer_id }
}

export function setActiveDisplayMode(canvas: Canvas, mode_id: string): Canvas {
  if (!canvas.display_modes.find((m) => m.id === mode_id)) throw new Error(`display mode ${mode_id} not found`)
  return { ...canvas, active_display_mode: mode_id }
}

export function getLayer(canvas: Canvas, layer_id: string): CanvasLayer | null {
  return canvas.layers.find((l) => l.id === layer_id) ?? null
}

/**
 * Validate a canvas object.
 */
export function validateCanvas(canvas: unknown): { ok: boolean; errors: string[] } {
  const errors: string[] = []
  if (!canvas || typeof canvas !== 'object') {
    return { ok: false, errors: ['canvas must be an object'] }
  }
  const c = canvas as Partial<Canvas> & Record<string, unknown>
  if (c.version !== 1) errors.push('version must be 1')
  if (!Array.isArray(c.layers)) {
    errors.push('layers must be an array')
  } else {
    const ids = new Set()
    c.layers.forEach((l: Partial<CanvasLayer>, i: number) => {
      if (!l.id)   errors.push(`layers[${i}] missing id`)
      if (!l.name) errors.push(`layers[${i}] missing name`)
      if (l.color && !HEX_RE.test(l.color)) errors.push(`layers[${i}] invalid color: ${l.color}`)
      if (ids.has(l.id)) errors.push(`duplicate layer id: ${l.id}`)
      ids.add(l.id)
    })
    if (!c.layers.find((l: Partial<CanvasLayer>) => l.id === c.active_layer)) {
      errors.push(`active_layer '${c.active_layer}' not in layers`)
    }
  }
  if (!Array.isArray(c.display_modes)) {
    errors.push('display_modes must be an array')
  } else {
    if (!c.display_modes.find((m: Partial<CanvasDisplayMode>) => m.id === c.active_display_mode)) {
      errors.push(`active_display_mode '${c.active_display_mode}' not in display_modes`)
    }
  }
  return { ok: errors.length === 0, errors }
}
